Document pokecenter common scripts and animation

metallicity opened the pull request:
Documented a few common scripts all related to healing the party, both when visiting a pokecenter and also after losing a battle (with separate scripts for landing at a pokecenter or at home). Along with this, I also documented the animation used for the pokecenter healing machine, which has a near identical copy (minus the flashing screen and sound effects) when entering a player into the hall of fame. This is the second such pair of animations I've encountered with extremely similar implementation, which does lead me to wonder if Game Freak just loves copy/pasting or had some sort of shared build process that produced the logic for both these animations.

I've also avoided the question of how to reference common scripts in this PR, instead just adding a few constants for the exact common scripts that were documented. See draft PR #687 for my current ideas of how this could be managed for all common scripts. I did rename `Common_` to `CommonScript_`, since I felt that would be a better prefix to use consistently both inside and outside this file, but don't feel too strongly about it. It just seemed best to consider a change like this now, when only a handful of common scripts are documented and/or referenced by name.

scbroede commented on src/trainer_card_screen/trainer_card_screen.c:
```diff
@@ -483,23 +484,23 @@ static void TrainerCard_LoadCardPalette(u8 stars, u8 pokedexObtained, NARC *narc
-        case 0:
-            nclrBuffer = Graphics_GetPlttDataFromOpenNARC(narc, 0, &paletteData, HEAP_ID_TRAINER_CARD_SCREEN);
+        case TRAINER_CARD_LEVEL_NORMAL:
+            nclrBuffer = Graphics_GetPlttDataFromOpenNARC(narc, TRAINER_CARD_LEVEL_NORMAL, &paletteData, HEAP_ID_TRAINER_CARD_SCREEN);
```
the narc member indices should have their own constants (ie. something like trainer_card_NCLR). but leave these as numbers for now, those constants will be generated once the relevant narc is unpacked

lhearachel replied:
+1 here; just keep the raw values for now; feel free to leave comments denoting what they are, as breadcrumbs for future contribution.

scbroede commented on src/trainer_card_screen/trainer_card_screen.c:
```diff
@@ -483,23 +484,23 @@ static void TrainerCard_LoadCardPalette(u8 stars, u8 pokedexObtained, NARC *narc
 
     if (pokedexObtained) {
         switch (stars) {
```
the `stars` variable should probably be changed to `trainerCardLevel` or something similar to match the constant names

lhearachel replied:
This seems like a note for a future PR that revisits this screen.

metallicity replied:
It's also possible to change the constant names to match "stars". It's just that with two possible names for each constant, and one of them would just append each number to some prefix, I leaned towards the descriptive, non-numerical names instead, since the former is already implicitly documented by the values. I do admit thought that in the case of a strictly incrementing constant like this, whose values are also used in less-than/greater-than operations, it seems more reasonable to put the numerical values more up front.

lhearachel replied:
I think we can just leave them for now until we unpack the NARC itself, since we're just keeping the raw literals here. Is that agreeable?

lhearachel replied:
I'm realizing now that the constants and the NARC files are implicitly coupled; do we want to just trash the `constants` added in this PR and revisit them later when we unpack the NARC...?

metallicity replied:
There is an extra "doesn't have the pokedex" palette index `6` for this narc, which is not a possible result from `TrainerCard_CalculateStars`, for however much that matters to you. Loading up a new save file, the only difference is that the area for the pokedex count is omitted, but this does mean that trainer card levels and trainer card palettes do not have a 1:1 relationship.

More broadly though, I wouldn't expect a constant showing up in `scripts_common.s` to be called something like `TRAINER_CARD_PALETTE_INDEX_GOLD`, when it isn't dealing with palettes at all. Having two sets of constants or the palette indexes borrowing the corresponding `TRAINER_CARD_LEVEL_GOLD` (or `TRAINER_CARD_STARS_4`, or what have you) both seem more natural options.

Here are some screenshots for comparison, though the "Pokedex" text and count are missing since the pokedex isn't actually acquired yet.

### Normal Trainer Card (Without Pokedex) Palette

<img width="611" height="451" alt="trainer_card_normal_no_pokedex" src="https://github.com/user-attachments/assets/67c7b7c4-2af6-4b2d-a654-0a091f16cefe" />

### Normal Trainer Card (With Pokedex) Palette

<img width="611" height="451" alt="trainer_card_normal_pokedex" src="https://github.com/user-attachments/assets/d8164fac-ec89-4b4b-8adb-bd499728755f" />

lhearachel commented on src/unk_020528D0.c:
```diff
@@ -245,11 +244,11 @@ BOOL sub_02052B2C(FieldTask *task)
         BrightnessController_SetScreenBrightness(0, GX_BLEND_PLANEMASK_BG0 | GX_BLEND_PLANEMASK_BG1 | GX_BLEND_PLANEMASK_BG2 | GX_BLEND_PLANEMASK_BG3 | GX_BLEND_PLANEMASK_OBJ | GX_BLEND_PLANEMASK_BD, BRIGHTNESS_BOTH_SCREENS);
 
-        if (sub_0203A7EC()
+        if (FieldOverworldState_GetDefaultWarpID()
             == FieldOverworldState_GetWarpId(SaveData_GetFieldOverworldState(fieldSystem->saveData))) {
```
**polish**: Pull the right-hand-side up to the previous line. Splitting the equality-operator is a bit confusing.

metallicity commented:
@lhearachel I went ahead and removed all the script constants here, and just left them undocumented in `CallCommonScript` and the like for now. Shouldn't be too hard to track them down again once we actually have the constants ready, since there aren't many too literals around the range of `0x7D0`. I also rebased on the latest version of main, since the merge conflicts were starting to pile up a bit. Let me know if there's anything else that still needs to be addressed before this can be merged.

lhearachel reviewed (approved):
One more piece of polish-feedback and then fix merge conflicts.

metallicity commented on generated/player_request_states.txt:
Same. I think the issue might be that `PlayerAvatar_RequestState[PlayerState]` was instead interpreted as `PlayerAvatar_RequestState[PlayerRequestState]`, since there were a bunch more states than the 3 possible values `PlayerState` appeared to have (when it could be there are just 3 states specifically where the player is actionable, as all the additional documented states aren't). And then this was recently made more confusing by some (but not all) of these references being renamed to `TransitionState`. Which is maybe a little more clear when taken in isolation, but still not particularly so, especially now that `PlayerState`, `TransitionState` and `RequestState` are all floating around together probably referring to the exact same thing.

A good starting point for using `PLAYER_STATE_` constants everywhere might look something like:

- `playerAvatar->unk_04` -> `playerAvatar->next_state`
- `PlayerAvatar_SetRequestStateFlag` -> `PlayerAvatar_SetNextState`
- `PlayerAvatar_TurnOnRequestStateBit` -> `PlayerAvatar_SetNextStateFlagOn`
- `PlayerAvatar_RequestStateFlag` -> `PlayerAvatar_GetNextState`
- `ScrCmd_SetPlayerState` -> `ScrCmd_SetNextPlayerState`
- `ScrCmd_ChangeRequestState` -> `ScrCmd_TransitionPlayerState`

Possibly with a combined `ChangePlayerState` macro, since the two script commands are never called separately and there's no real reason to want to do so.

---

All that said, I'm just going to stick with `PLAYER_TRANSITION_STATE_` for this PR (and also fix the merge issue where these were unintentionally split into separate sets of constants). That's what these constants are currently called in `main`, and these changes just want to refer to one of those constants in a script.

metallicity replied:
Actually, looking at what it's doing more closely, it seems like it might be more of a "temporary state" than an actual transition. The player's original state (walking/cycling/surfing) is retained, and all the places I see it used eventually return the player to this state.

metallicity commented:
@narf0708 I put our `scripts_common.s` changes side by side and took on some of your terminology in places, just to ease the process of landing on a single set of names. The base prefix I went with was `CommonScript_PokecenterNurse`, which is more or less what was discussed on Discord, just without the additional underscore for `Pokecenter_Nurse`, which felt a little extraneous. I think all (or at least nearly all) script entries under `CommonScript_SomeName` should aim have just a single phrase after the prefix, so that their own sub-labels can use `CommonScript_SomeName_` as a prefix without things getting too unwieldy.

Leave some review comments if there are any names/naming patterns here you'd still want to see changed.

metallicity commented:
@lhearachel Rebased this with the changes added to main by #688. Should be fine to merge in now.
